### PR TITLE
[extractor/nhk] add NHK Radiru Live IE

### PR DESCRIPTION
**Pull request by garret1317**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

This PR adds an extractor for NHK Radiru's live radio streams.
It supports the regional variations by letting them be specified as an extractor arg. For example, Radio 1 in Fukuoka would be `yt-dlp https://www.nhk.or.jp/radio/player/?ch=r1 --extractor-args nhkradirulive:area=fukuoka`.
If no area is specified, it defaults to tokyo. If an invalid area is specified, an error is thrown that lists the valid areas.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
📻🇯🇵🆕

Add support for NHK Radiru Live, a Japanese radio streaming service, by creating a new extractor class `NhkRadiruLiveIE` and registering it in `_extractors.py`.

> _`NhkRadiruLiveIE`_
> _Extracts live radio streams_
> _Autumn leaves whisper_

### Walkthrough
*  Add `NhkRadiruLiveIE` extractor for NHK Radiru Live radio streaming service ([link](https://github.com/yt-dlp/yt-dlp/pull/7332/files?diff=unified&w=0#diff-780b22dc7eb280f5a7b2bbf79aff17826de88ddcbf2fc1116ba19901827aa4e3R1263), [link](https://github.com/yt-dlp/yt-dlp/pull/7332/files?diff=unified&w=0#diff-5adc25ce69895042565846073090a631920a91ba6326b97b2357a097a6c2c722R498-R568))
   * Register the extractor in `_extractors.py` ([link](https://github.com/yt-dlp/yt-dlp/pull/7332/files?diff=unified&w=0#diff-780b22dc7eb280f5a7b2bbf79aff17826de88ddcbf2fc1116ba19901827aa4e3R1263))
   * Define the extractor class in `nhk.py` ([link](https://github.com/yt-dlp/yt-dlp/pull/7332/files?diff=unified&w=0#diff-5adc25ce69895042565846073090a631920a91ba6326b97b2357a097a6c2c722R498-R568))
      * Import necessary modules and classes ([link](https://github.com/yt-dlp/yt-dlp/pull/7332/files?diff=unified&w=0#diff-5adc25ce69895042565846073090a631920a91ba6326b97b2357a097a6c2c722L10-R13))
      * Specify geo-restriction, URL pattern, and test cases ([link](https://github.com/yt-dlp/yt-dlp/pull/7332/files?diff=unified&w=0#diff-5adc25ce69895042565846073090a631920a91ba6326b97b2357a097a6c2c722R498-R568))
      * Implement `_real_extract` method to download and parse data from website and API ([link](https://github.com/yt-dlp/yt-dlp/pull/7332/files?diff=unified&w=0#diff-5adc25ce69895042565846073090a631920a91ba6326b97b2357a097a6c2c722R498-R568))

</details>

### CODE REVIEW

File: yt_dlp/extractor/nhk.py [hunk 1]
```diff
@@ -492,3 +495,74 @@ class NhkRadioNewsPageIE(InfoExtractor):
...
+    }]
+
+    _NOA_STATION_IDS = {'r1': 'n1', 'r2': 'n2', 'fm': 'n3'}  # lookup table is easier than working it out
+
+    def _real_extract(self, url):
+        station = self._match_id(url)
+        area = traverse_obj(self._configuration_arg('area'), 0) or "tokyo"
+
+        config = self._download_xml('https://www.nhk.or.jp/radio/config/config_web.xml', None, 'Downloading area information')
+        data = config.find(f'.//data//area[.="{area}"]/..')  # a <data> element with an <area> that matches the ^variable
```
Comment (bashonly) [inline yt_dlp/extractor/nhk.py:545]: first line too long, you can remove comment
```suggestion
        config = self._download_xml(
            'https://www.nhk.or.jp/radio/config/config_web.xml', station, 'Downloading area information')
        data = config.find(f'.//data//area[.="{area}"]/..')
```

File: yt_dlp/extractor/nhk.py [hunk 2]
```diff
@@ -492,3 +495,74 @@ class NhkRadioNewsPageIE(InfoExtractor):
...
+
+        if data is None:
+            raise ExtractorError('Invalid area - valid areas are: %s'
+                                 % ', '.join([i.text for i in config.findall('.//data//area')]),
+                                 expected=True)
```
Comment (bashonly) [inline yt_dlp/extractor/nhk.py:550]:
```suggestion
        if not data:
            raise ExtractorError('Invalid area. Valid areas are: %s' % ', '.join(
                [i.text for i in config.findall('.//data//area')]), expected=True)
```

File: yt_dlp/extractor/nhk.py [hunk 3]
```diff
@@ -492,3 +495,74 @@ class NhkRadioNewsPageIE(InfoExtractor):
...
+
+        noa_info = self._download_json(f'https:{config.find(".//url_program_noa").text}'
+                                       .replace("{area}", data.find('areakey').text), station)
+
+        present_info = traverse_obj(noa_info, ('nowonair_list', self._NOA_STATION_IDS.get(station), 'present'))
```
Comment (bashonly) [inline yt_dlp/extractor/nhk.py:555]: `.format()` is a bit cleaner than the replace imo
```suggestion
        noa_info = self._download_json(
            f'https:{config.find(".//url_program_noa").text}'.format(area=data.find('areakey').text), station)
        present_info = traverse_obj(noa_info, ('nowonair_list', self._NOA_STATION_IDS.get(station), 'present'))
```

File: yt_dlp/extractor/nhk.py [hunk 4]
```diff
@@ -492,3 +495,74 @@ class NhkRadioNewsPageIE(InfoExtractor):
...
+        service = present_info.get('service')
+
+        return {
+            'title': join_nonempty(service.get('name'), traverse_obj(present_info, ('area', 'name')), delim=" "),
+            'id': join_nonempty(station, area),
+            'thumbnails': traverse_obj(service, ('images', ..., {
```
Comment (bashonly) [inline yt_dlp/extractor/nhk.py:561]: we can simplify `title` extraction with a single traversal instead, and eliminate the need for the `service` var
```suggestion

        return {
            'title': ' '.join(traverse_obj(present_info, (('service', 'area',), 'name', {str}))),
            'id': join_nonempty(station, area),
            'thumbnails': traverse_obj(present_info, ('service', 'images', ..., {
```
also, is `join_nonempty` needed for `station` and `area`? will it ever be the case that one of them is empty?

File: yt_dlp/extractor/nhk.py [hunk 5]
```diff
@@ -492,3 +495,74 @@ class NhkRadioNewsPageIE(InfoExtractor):
...
+                                       'url': 'url',
+                                       'width': ('width', {int_or_none}),
+                                       'height': ('height', {int_or_none}),
+                                       })),
```
Comment (bashonly) [inline yt_dlp/extractor/nhk.py:565]: I prefer this formatting/indent, but I guess either is fine
```suggestion
                'url': 'url',
                'width': ('width', {int_or_none}),
                'height': ('height', {int_or_none}),
            })),
```

File: yt_dlp/extractor/nhk.py [hunk 6]
```diff
@@ -7,7 +7,10 @@
     unescapeHTML,
     unified_timestamp,
     urljoin,
-    url_or_none
+    url_or_none,
+    join_nonempty,
+    int_or_none,
+    ExtractorError,
```
Comment (bashonly) [inline yt_dlp/extractor/nhk.py:13]: sort these: classes then functions, then sort alphabetically
or just use `isort`

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/nhk.py [hunk 7]
```diff
@@ -492,3 +495,74 @@ class NhkRadioNewsPageIE(InfoExtractor):
...
+            'thumbnail': 'https://www.nhk.or.jp/common/img/media/fm-200x200.png',
+            'live_status': 'is_live',
+        }
+    }]
+
+    _NOA_STATION_IDS = {'r1': 'n1', 'r2': 'n2', 'fm': 'n3'}  # lookup table is easier than working it out
+
+    def _real_extract(self, url):
+        station = self._match_id(url)
+        area = traverse_obj(self._configuration_arg('area'), 0) or "tokyo"
```
[the remark this answers is not in the document]
Comment (garret1317) [reply yt_dlp/extractor/nhk.py:542]: > but how does the website handle this?

as far as i can tell, there's just a dropdown with the areas that sets a cookie
on the homepage:
the dropdown in question, with the firefox dev tools showing a cookie of _rr_Area: tokyo

on the player page:
another dropdown, this one on the player page
neither of them affect the url.

> does it determine which area to serve the user based on IP/location?

dont think so, it defaults to tokyo even from my uk IP.
tried with mullvad osaka, that defaulted to tokyo as well (though that may not mean anything, i've had radiko detect the osaka servers as tokyo before)
> there's nothing in the URL/query?

unfortunately not, seems to be just the channel id

File: yt_dlp/extractor/nhk.py (hunk 5 above)
Comment (garret1317) [reply yt_dlp/extractor/nhk.py:565]: yeah that's fine, i thought i had to do it the way i did for flake8, but actually it's fine with either

File: yt_dlp/extractor/nhk.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (bashonly) [reply yt_dlp/extractor/nhk.py:542]: extractor-arg works then

File: yt_dlp/extractor/nhk.py (hunk 3 above)
Comment (bashonly) [reply yt_dlp/extractor/nhk.py:555]: add a note to the `noa_info` `_download_json` about which `area` is being downloaded

File: yt_dlp/extractor/nhk.py (hunk 4 above)
Comment (garret1317) [reply yt_dlp/extractor/nhk.py:561]: > will it ever be the case that `station` or `area` is empty?

probably not, it's just more convenient to use the nice util function imo

**Comment by garret1317** [comment]
unrelated to this PR, but while writing the section in the readme i noticed that `wrestleuniverse` seems to be off by one heading level (h3 rather than h4)
Screenshot from 2023-06-18 23-25-51

**Comment by bashonly** [comment]
@garret1317 I noticed that too. Was going to suggest it for the next cleanup commit

File: yt_dlp/extractor/nhk.py [hunk 8]
```diff
@@ -492,3 +495,73 @@ class NhkRadioNewsPageIE(InfoExtractor):
...
+        'info_dict': {
+            'id': 'fm-sapporo',
+            'title': 're:^ＮＨＫネットラジオＦＭ 札幌.+$',
+            'ext': 'm4a',
+            'thumbnail': 'https://www.nhk.or.jp/common/img/media/fm-200x200.png',
+            'live_status': 'is_live',
+        }
+    }]
+
+    _NOA_STATION_IDS = {'r1': 'n1', 'r2': 'n2', 'fm': 'n3'}  # lookup table is easier than working it out
```
Comment (bashonly) [inline yt_dlp/extractor/nhk.py:538]: agreed re: comment, but we can cleanup
```suggestion
    _NOA_STATION_IDS = {'r1': 'n1', 'r2': 'n2', 'fm': 'n3'}
```

**Review by bashonly — APPROVED** [review]

**Pull request merged**